urbit/app/eth-event-viewer.hoon

```diff
@@ -0,0 +1,409 @@
+::
+++  event-logs-card
+  |=  event-logs=loglist:ew-sur
+  ^-  (quip card _state)
+  ?~  event-logs
+    [~ state]
+  =/  address
+    address:`event-log:rpc:ethereum`+2:event-logs
+  =/  contract
+    (~(got by contracts.state) address)
+  =/  updated-contract
+    contract(event-logs event-logs)
+  =/  filtered-contracts
+    (~(del by contracts.state) address)
+  =/  new-contracts
+    (~(put by filtered-contracts) address updated-contract)
+  :_  state(contracts new-contracts)
+  [%give %fact [/state/update ~] %json !>((history-json:ew-lib event-logs))]~
+::
+++  event-log-card
+  |=  =event-log:rpc:ethereum
+  ^-  (quip card _state)
+  =/  address
+    address.event-log
+  =/  contract
+    (~(got by contracts.state) address)
+  =/  updated-contract
+    contract(event-logs (weld (flop event-logs.contract) [event-log ~]))
+  =/  filtered-contracts
+    (~(del by contracts.state) address)
+  =/  new-contracts
+    (~(put by filtered-contracts) address updated-contract)
+  :_  state(contracts new-contracts)
+  [%give %fact [/state/update ~] %json !>((event-log-json:ew-lib event-log))]~
+::
+++  request-ethereum-abi
+  |=  address=@ux
+  ^-  card:agent:gall
+  =/  req=request:http
+    (get-request (ux-to-cord address))
+  [%pass /eth-event-viewer/abi-res %arvo %i %request req *outbound-config:iris]
+::
+++  get-request
+  |=  address=@t
+  ^-  request:http
+  =/  base=@t
+    'https://api.etherscan.io/api?module=contract&action=getabi&address='
+  =/  url=@t  (cat 3 base address)
+  =/  hed  [['Accept' 'application/json']]~
+  [%'GET' url hed *(unit octs)]
+::
+++  ux-to-cord
+  |=  address=@ux
+  ^-  @t
+::  crip transforms tape to cord
+  (cat 3 '0x' (crip ((x-co:co 4) address)))
+::
+++  to-eth-watcher
+  |=  [=wire =task:agent:gall]
+  ^-  card
+  [%pass wire %agent [our.bol %eth-watcher] task]
+::
+++  setup-eth-watcher
+  |=  contract=contract-type
+  =/  url  'http://eth-mainnet.urbit.org:8545'
+  %+  to-eth-watcher  (get-path address.contract)
+  :+  %poke   %eth-watcher-poke
+  !>  ^-  poke:ew-sur
+  :+  %watch  (get-path address.contract)
+  :*  url
+      |
+      ~m2
+      ~m2
+      9.960.593
+      ~[address.contract]
+      (get-topics specific-events.contract)
+  ==
+::
+++  watch-eth-watcher
+  |=  address=@ux
+  %+  to-eth-watcher  (get-path address)
+  [%watch (get-logs-path address)]
+::
+++  leave-eth-watcher
+  |=  address=@ux
+  %+  to-eth-watcher  (get-path address)
+  [%leave ~]
+::
+++  clear-eth-watcher
+  |=  address=@ux
+  %+  to-eth-watcher  /clear
+  :+  %poke  %eth-watcher-poke
+  !>  ^-  poke:ew-sur
+  [%clear (get-path address)]
+::
+++  get-topics
+  |=  specific-events=(list @t)
+  ^-  (list ?(@ux (list @ux)))
+  ?~  specific-events
+    ~
+  :~  `(list @ux)`(turn specific-events hex-list)
+  ==
+::
+++  hex-list  |=(e=@t `@ux`(event-string-to-hex (trip e)))
+::
+++  event-string-to-hex
+  |=  event-string=tape
+  ^-  @ux
+  `@ux`(keccak-256:keccak:crypto (as-octt:mimes:html event-string))
```

Comment:
So why doesn't `+event-string-to-hex` just take a `@t`, and do the `+trip` itself? (;  
It's not called anywhere else I don't think.

And then `+hex-list`, also, is only called in one place. You could write the above as something like

```hoon
?~  specific-events  ~
:_  ~
%+  turn  specific-events
:(cork trip as-octt:mimes:html keccak-256:keccak:crypto)
```